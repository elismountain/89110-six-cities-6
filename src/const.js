export const Cities = {
  PARIS: `Paris`,
  COLOGNE: `Cologne`,
  BRUSSELS: `Brussels`,
  AMSTERDAM: `Amsterdam`,
  HAMBURG: `Hamburg`,
  DUSSELDORF: `Dusseldorf`
};


export const cardTypes = {
  MAIN: `main`,
  FAVORITES: `favorites`,
  NEARBY: `nearby`
};

export const currentCity = Cities.AMSTERDAM;

export const RATING = [
  {
    value: 5,
    title: `perfect`
  },
  {
    value: 4,
    title: `good`
  },
  {
    value: 3,
    title: `not bad`
  },
  {
    value: 2,
    title: `bad`
  },
  {
    value: 1,
    title: `terrible`
  }
];

export const OfferTypes = {
  APARTMENT: `Apartment`,
  ROOM: `Private room`,
  HOUSE: `House`,
  HOTEL: `Hotel`
};
