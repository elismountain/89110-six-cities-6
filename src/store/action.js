export const ActionType = {
  REQUIRED_AUTHORIZATION: `user/requiredAuthorization`,
  SET_AUTH_INFO: `user/login`,
  CHANGE_CITY: `main/changeCity`,
  CHANGE_SORTING: `main/changeSorting`,
  LOAD_OFFERS: `data/loadOffers`,
  LOAD_OFFER: `offer/loadDetails`,
  LOAD_REVIEWS: `offer/loadReviews`,
  LOAD_NEARBY: `offer/loadNearby`,
  OFFER_NOT_FOUND: `offer/notFound`,
  RESET_OFFER_STATE: `offer/resetState`,
  ADD_FAVORITE: `data/addFavorite`,
  REMOVE_FAVORITE: `data/removeFavorite`,
  LOAD_FAVORITES: `data/loadFavorites`,
  RESET_FAVORITES: `data/resetFavorites`,
  REDIRECT_TO_ROUTE: `middlewares/redirectToRoute`
};

export const ActionCreator = {
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: status,
  }),
  setAuthInfo: (info) => ({
    type: ActionType.SET_AUTH_INFO,
    payload: info
  }),
  loadOffers: (offers) => ({
    type: ActionType.LOAD_OFFERS,
    payload: offers
  }),

  loadOfferById: (offer) => ({
    type: ActionType.LOAD_OFFER,
    payload: offer,
  }),

  loadReviews: (reviews) => ({
    type: ActionType.LOAD_REVIEWS,
    payload: reviews
  }),

  loadNearby: (offers) => ({
    type: ActionType.LOAD_NEARBY,
    payload: offers
  }),

  loadFavorites: (offers) => ({
    type: ActionType.LOAD_FAVORITES,
    payload: offers,
  }),

  offerNotFound: () => ({
    type: ActionType.OFFER_NOT_FOUND
  }),

  resetOfferState: () => ({
    type: ActionType.RESET_OFFER_STATE
  }),

  changeCity: (city) => ({
    type: ActionType.CHANGE_CITY,
    payload: city,
  }),

  changeSorting: (sorting) => ({
    type: ActionType.CHANGE_SORTING,
    payload: sorting
  }),

  addFavorite: (offer) => ({
    type: ActionType.ADD_FAVORITE,
    payload: offer
  }),

  removeFavorite: (offer) => ({
    type: ActionType.REMOVE_FAVORITE,
    payload: offer
  }),

  resetFavorites: () => ({
    type: ActionType.RESET_FAVORITES,
  }),

  redirectToRoute: (url) => ({
    type: ActionType.REDIRECT_TO_ROUTE,
    payload: url
  })
};
