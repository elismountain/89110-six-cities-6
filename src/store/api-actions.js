import {ActionCreator} from './action';
import {AppRoutes, APIRoutes, AuthorizationStatus} from '../const';
import {ActionCreator as middlewareActionCreator} from './middlewares/action';

export const checkAuth = () => (dispatch, _getState, api) => (
  api.get(APIRoutes.LOGIN)
    .then(({data}) => {
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      dispatch(ActionCreator.setAuthInfo(data));
    })
    .catch(() => {})
);

export const login = ({email, password}) => (dispatch, _getState, api) => (
  api.post(APIRoutes.LOGIN, {email, password})
    .then(({data}) => {
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      dispatch(ActionCreator.setAuthInfo(data));
    })
    .then(() => dispatch(middlewareActionCreator.redirectToRoute(AppRoutes.MAIN)))
    .catch(() => {})
);

export const logout = () => (dispatch, _getState, api) => (
  api.get(APIRoutes.LOGOUT)
    .then(() => {
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.NO_AUTH));
      dispatch(ActionCreator.setAuthInfo({}));
      dispatch(ActionCreator.resetFavorites());
    })
    .catch(() => {})
);

export const fetchOffers = () => (dispatch, _getState, api) => (
  api.get(APIRoutes.OFFERS)
    .then(({data}) => dispatch(ActionCreator.loadOffers(data)))
);

export const fetchOfferById = (id) => (dispatch, _getState, api) => (
  api.get(`/hotels/${id}`)
    .then(({data}) => dispatch(ActionCreator.loadOfferById(data)))
    .catch(() => dispatch(ActionCreator.offerNotFound()))
);

export const fetchReviews = (id) => (dispatch, _getState, api) => (
  api.get(`/comments/${id}`)
    .then(({data}) => dispatch(ActionCreator.loadReviews(data)))
    .catch(() => {})
);

export const fetchNearby = (id) => (dispatch, _getState, api) => (
  api.get(`/hotels/${id}/nearby`)
    .then(({data}) => dispatch(ActionCreator.loadNearby(data)))
    .catch(() => {})
);

export const sendReview = ({id, review}) => (dispatch, _getState, api) => (
  api.post(`/comments/${id}`, review)
    .then(({data}) => dispatch(ActionCreator.loadReviews(data)))
);

export const fetchFavorites = () => (dispatch, _getState, api) => (
  api.get(APIRoutes.FAVORITES)
    .then(({data}) => dispatch(ActionCreator.loadFavorites(data)))
);

export const changeOfferStatus = (id, status) => (dispatch, _getState, api) => (
  api.post(`/favorite/${id}/${status}`)
    .then(({data}) => {
      dispatch(ActionCreator[data[`is_favorite`] ? `addFavorite` : `removeFavorite`](data));
    })
);
