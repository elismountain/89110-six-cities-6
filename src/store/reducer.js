import {ActionType} from './action';
import {Cities, SortingTypes} from '../const';
import {getCityOffers, getFavoriteOffers} from '../utils';
import offers from '../mocks/offers';

const initialState = {
  activeCity: Cities.PARIS,
  offers,
  cityOffers: getCityOffers(offers, Cities.PARIS),
  favoriteOffers: getFavoriteOffers(offers),
  activeSorting: SortingTypes.POPULAR
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_CITY:
      return {
        ...state,
        activeCity: action.payload,
        cityOffers: getCityOffers(state.offers, state.activeCity)
      };
    case ActionType.SET_CITY_OFFERS:
      return {
        ...state,

      };
    case ActionType.CHANGE_SORTING:
      return {
        ...state,
        activeSorting: action.payload
      };
    default:
      return state;
  }
};

export {reducer};
