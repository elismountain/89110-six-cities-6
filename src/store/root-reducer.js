import {combineReducers} from 'redux';
import {reducer as mainReducer} from './main/reducer';
import {reducer as userReducer} from './user/reducer';
import {reducer as offerReducer} from './offer/reducer';
import {reducer as favoritesReducer} from './favorites/reducer';

export const NameSpace = {
  USER: `USER`,
  MAIN: `MAIN`,
  OFFER: `OFFER`,
  FAVORITES: `FAVORITES`
};

const rootReducer = combineReducers({
  [NameSpace.USER]: userReducer,
  [NameSpace.MAIN]: mainReducer,
  [NameSpace.OFFER]: offerReducer,
  [NameSpace.FAVORITES]: favoritesReducer
});

export {rootReducer};
