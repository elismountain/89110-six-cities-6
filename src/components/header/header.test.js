import React from 'react';
import {render} from '@testing-library/react';
import {Router} from 'react-router-dom';
import {Provider} from 'react-redux';
import {createMemoryHistory} from 'history';
import configureStore from 'redux-mock-store';
import Header from './header';
import {AuthorizationStatus} from '../../const';
import {NameSpace} from '../../store/root-reducer';

const mockStore = configureStore({});

describe(`Header should render correctly`, () => {
  it(`Header should render correctly with unauthorized user`, () => {
    const history = createMemoryHistory();
    const {container} = render(
        <Provider store={mockStore({
          [NameSpace.USER]: {
            authorizationStatus: AuthorizationStatus.NO_AUTH,
            authInfo: {},
          }
        })}>
          <Router history={history}>
            <Header />
          </Router>
        </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it(`Header should render correctly with authorized user`, () => {
    const history = createMemoryHistory();
    const {container} = render(
        <Provider store={mockStore({
          [NameSpace.USER]: {
            authorizationStatus: AuthorizationStatus.AUTH,
            authInfo: {email: `test@gmail.com`, name: `test`},
          }
        })}>
          <Router history={history}>
            <Header />
          </Router>
        </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
