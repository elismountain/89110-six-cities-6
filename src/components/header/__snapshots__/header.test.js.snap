// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly Header should render correctly with authorized user 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="container"
    >
      <div
        class="header__wrapper"
      >
        <div
          class="header__left"
        >
          <a
            class="header__logo-link header__logo-link--active"
            href="/"
          >
            <img
              alt="6 cities logo"
              class="header__logo"
              height="41"
              src="img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          class="header__nav"
        >
          <ul
            class="header__nav-list"
          >
            <li
              class="header__nav-item user"
            >
              <a
                class="header__nav-link header__nav-link--profile"
                href="/favorites"
              >
                <div
                  class="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  class="header__user-name user__name"
                >
                  test@gmail.com
                </span>
              </a>
            </li>
            <li
              class="header__nav-item"
              style="margin-left: 10px;"
            >
              <button
                class="button"
              >
                <svg
                  fill="none"
                  height="20"
                  stroke="#777"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
                  />
                  <polyline
                    points="16 17 21 12 16 7"
                  />
                  <line
                    x1="21"
                    x2="9"
                    y1="12"
                    y2="12"
                  />
                </svg>
              </button>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header should render correctly Header should render correctly with unauthorized user 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="container"
    >
      <div
        class="header__wrapper"
      >
        <div
          class="header__left"
        >
          <a
            class="header__logo-link header__logo-link--active"
            href="/"
          >
            <img
              alt="6 cities logo"
              class="header__logo"
              height="41"
              src="img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          class="header__nav"
        >
          <ul
            class="header__nav-list"
          >
            <li
              class="header__nav-item user"
            >
              <a
                class="header__nav-link header__nav-link--profile"
                href="/login"
              >
                <div
                  class="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  class="header__login"
                >
                  Sign in
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
</div>
`;
